File: test_suite/feature_tests/C/test_shmem_collects.c
/* (c) 2011 University of Houston System.  All rights reserved. */
/* Tests shmem_fcollect32 call
 * Each PE contributes 4 elements 
 * */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

#include <mpp/shmem.h>

static long pSync[_SHMEM_BCAST_SYNC_SIZE];

static int src1[4] = { 11, 12, 13, 14 };
static long src2[4] = { 101, 102, 103, 104 };
static int src3[4] = { 1, 2, 3, 4 };
static long src4[4] = { 11, 12, 13, 14 };


int npes;
int me;

volatile int x = 1;

int
main(void)
{
  int i,j,k,success;
  int dst_len,x,y;
  int *dst1, *dst3;
  long *dst2, *dst4;
  int *compare_dst1, *compare_dst3;
  long *compare_dst2, *compare_dst4;

  start_pes(0);
  npes = _num_pes();
  me = _my_pe();

  for (i = 0; i < _SHMEM_BCAST_SYNC_SIZE; i += 1) {
    pSync[i] = _SHMEM_SYNC_VALUE;
  }

  if(npes>1){


    dst1 = (int *) shmalloc(npes*4*sizeof(dst1));
    compare_dst1 = (int *) malloc(npes*4*sizeof(dst1));
    dst2 = (long *) shmalloc(npes*4*sizeof(dst2));
    compare_dst2 = (long *) malloc(npes*4*sizeof(dst2));
    dst3 = (int *) shmalloc(npes*4*sizeof(dst3));
    compare_dst3 = (int *) malloc(npes*4*sizeof(dst3));
    dst4 = (long *) shmalloc(npes*4*sizeof(dst4));
    compare_dst4 = (long *) malloc(npes*4*sizeof(dst4));

    /*Test shmem_fcollect32 */
    /*Create the output of fcollect32 and save in compare_dst array*/
    success = 0;
    j=0;
    k=0;
    for(i=0;i<npes;i++){
      for(j=0;j<4;j++){
        compare_dst1[k]= src1[j];
        k++;
      }
    }

    for (i = 0; i < (npes*4); i++) {
      dst1[i] = -1;
    }

    shmem_barrier_all();

    shmem_fcollect32(dst1, src1, 4, 0, 0, npes,
        pSync);


    if(me == 0){
      for (i = 0; i < npes*4; i+= 1) {
        if(dst1[i] != compare_dst1[i])
          success =1;
      }
      if(success==1)
        printf("Test shmem_fcollect32: Failed\n");
      else
        printf("Test shmem_fcollect32: Passed\n");
    }

    shmem_barrier_all();


    /*Create the output of fcollect64 and save in compare_dest array*/
    success=0;
    j=0;
    k=0;
    for(i=0;i<npes;i++){
      for(j=0;j<4;j++){
        compare_dst2[k]= src2[j];
        k++;
      }
    }

    for (i = 0; i < (npes*4); i++) {
      dst2[i] = -1;
    }

    shmem_barrier_all();

    shmem_fcollect64(dst2, src2, 4,
        0, 0, npes,
        pSync);


    if(me == 0){
      for (i = 0; i < npes*4; i+= 1) {
        if(dst2[i] != compare_dst2[i])
          success =1;
      }
      if(success==1)
        printf("Test shmem_fcollect64: Failed\n");
      else
        printf("Test shmem_fcollect64: Passed\n");
    }

    /*Test collect32*/
    success=0;
    /*Decide the length of the destination array*/
    x = npes/4;
    y = npes%4;

    if(y == 1)
      dst_len = x * 10 + 1;
    else if(y == 2)
      dst_len = x * 10 + 3;
    else if(y == 3)
      dst_len = x * 10 + 6;
    else
      dst_len = x * 10;

    /*Create the output of collect32 and save in compare_dst array*/

    j=0;
    k=0;
    for(i=0;i<npes;i++){
      for(j=0;j<(i%4 +1);j++){
        compare_dst3[k]=  i*10 + src3[j];
        /*printf("compare_dst[%d] = %d \n",k,compare_dst[k]);*/
        k++;
      }
    }


    for (i = 0; i < dst_len; i++) {
      dst3[i] = -1;
    }

    for(i=0;i<4;i++)
      src3[i]= me*10 + src3[i];

    shmem_barrier_all();

    shmem_collect32(dst3, src3, (me%4 + 1),
        0, 0, npes,
        pSync);



    if(me == 0){
      for (i = 0; i < dst_len; i+= 1) {
        if(dst3[i] != compare_dst3[i])
          success =1;
      }
      if(success==1)
        printf("Test shmem_collect32: Failed\n");
      else
        printf("Test shmem_collect32: Passed\n");
    }

    shmem_barrier_all();

    /*Test shmem_collect64*/
    success=0;
    if(y == 1)
      dst_len = x * 10 + 1;
    else if(y == 2)
      dst_len = x * 10 + 3;
    else if(y == 3)
      dst_len = x * 10 + 6;
    else
      dst_len = x * 10;

    /*Create the output of collect64 and save in compare_dst array*/

    j=0;
    k=0;
    for(i=0;i<npes;i++){
      for(j=0;j<(i%4 +1);j++){
        compare_dst4[k]= src4[j];
        /*printf("compare_dst[%d]
         * =
         * %d
         * \n",k,compare_dst[k]);*/
        k++;
      }
    }


    for (i = 0; i < dst_len; i++) {
      dst4[i] = -1;
    }

    shmem_barrier_all();

    shmem_collect64(dst4, src4, (me%4 + 1),
        0, 0, npes,
        pSync);
    shmem_barrier_all();

    if(me == 0){
      for (i = 0; i < dst_len; i+= 1) {
        if(dst4[i] != compare_dst4[i])
          success =1;
      }
      if(success==1)
        printf("Test shmem_collect64: Failed\n");
      else
        printf("Test shmem_collect64: Passed\n");
    }

  }
  else
    printf("Number of PEs must be > 1 to test collects, test skipped\n");


  return 0;
}

File: test_suite/feature_tests/Fortran/test_shmem_collects.f90
! (c) 2011 University of Houston System.  All rights reserved. */
! Tests shmem_fcollect4/8 shmem_collect4/8 
! Each PE contributes 4 elements 
! 
program test_shmem_collects
  implicit none
  include 'mpp/shmem.fh'
 
  integer  , parameter :: dst_size = 8 ! assuming 2 pes ( 2 x 4 elements)

  integer,   save :: pSync(SHMEM_BCAST_SYNC_SIZE)

  integer*4, save :: src1(4) = (/ 11, 12, 13, 14 /) 
  integer*8, save :: src2(4) = (/ 101, 102, 103, 104 /)
  integer*4, save :: src3(4) = (/ 1, 2, 3, 4 /)
  integer*8, save :: src4(4) = (/ 11, 12, 13, 14 /)

  integer         :: npes
  integer         :: me

  integer,   save :: x

  integer         :: i, j, k, success
  integer         :: dst_len, y

  integer*4, save :: dst1(dst_size), dst3(dst_size) 
  integer*8, save :: dst2(dst_size), dst4(dst_size)
  integer*4, save :: compare_dst1(dst_size), compare_dst3(dst_size) 
  integer*8, save :: compare_dst2(dst_size), compare_dst4(dst_size)

  integer         :: length, errcode, abort

! Function definitions
  integer                   :: my_pe, num_pes
  
  x = 1

  call start_pes(0)
  npes = num_pes()
  me   = my_pe()

  do i = 1, SHMEM_BCAST_SYNC_SIZE, 1
    pSync(i) = SHMEM_SYNC_VALUE
  end do

  if(npes .gt. 1) then

    ! Test shmem_fcollect4
    ! Create the output of fcollect4 and save in compare_dst array
    success = 0
    k = 1
    do i = 1, npes, 1
      do j = 1, 4, 1
        compare_dst1(k) = src1(j)
        k = k + 1
      end do
    end do

    do i = 1, dst_size, 1
      dst1(i) = -1
    end do

    call shmem_barrier_all()

    call shmem_fcollect4(dst1, src1, 4, &
      0, 0, npes, &
      pSync)
    
    if(me .eq. 0) then
      do i = 1, npes * 4, 1
        if(dst1(i) .ne. compare_dst1(i)) then
          success = 1
        end if
      end do

      if(success .eq. 1) then
        write(*,*) "Test shmem_fcollect4: Failed"
      else
        write(*,*) "Test shmem_fcollect4: Passed"
      end if
    end if 

    call shmem_barrier_all()

    ! Create the output of fcollect8 and save in compare_dest array
    success = 0
    j = 0
    k = 1
    
    do i = 1, npes, 1
      do j = 1, 4, 1
        compare_dst2(k) = src2(j);
        k = k + 1
      end do
    end do

    do i = 1, npes * 4, 1
      dst2(i) = -1
    end do

    call shmem_barrier_all()

    call shmem_fcollect8(dst2, src2, 4, 0, 0, npes, pSync)

    if(me .eq. 0) then
      do i = 1, npes * 4, 1
        if(dst2(i) .ne. compare_dst2(i)) then
          success = 1
        end if
      end do

      if(success .eq. 1) then
        write(*,*) "Test shmem_fcollect8: Failed"
      else
        write(*,*) "Test shmem_fcollect8: Passed"
      end if
    end if

    ! Test collect32
    success = 0
    ! Decide the length of the destination array
    x = npes / 4
    y = mod(npes, 4)

    if(y .eq. 1) then
      dst_len = x * 10 + 1
    else if(y .eq. 2) then
      dst_len = x * 10 + 3
    else if(y .eq. 3) then
      dst_len = x * 10 + 6
    else
      dst_len = x * 10
    end if

    ! Create the output of collect32 and save in compare_dst array
    k = 1        
    
    do i = 0, npes - 1, 1
      do j = 1, mod(i, 4) + 1, 1
        compare_dst3(k)=  i * 10 + src3(j)
        k = k + 1
      end do
    end do
		
    do i = 1, dst_len, 1
      dst3(i) = -1
    end do

    do i = 1, 4, 1
      src3(i) = me * 10 + src3(i)
    end do
    
    call shmem_barrier_all()

    call shmem_collect4(dst3, src3, mod(me, 4) + 1, &
    	0, 0, npes, &
    	pSync)

    if(me .eq. 0) then
      
      do i = 1, dst_len, 1
        if(dst3(i) .ne. compare_dst3(i)) then
          success = 1
        end if
      end do

      if(success .eq. 1) then
        write (*,*) "Test shmem_collect4: Failed"
      else
        write (*,*) "Test shmem_collect4: Passed"
      end if
    end if 

    call shmem_barrier_all()

    ! Test shmem_collect8
    success = 0
    if(y .eq. 1) then
      dst_len = x * 10 + 1
    else if(y .eq. 2) then
      dst_len = x * 10 + 3
    else if(y .eq. 3) then
      dst_len = x * 10 + 6
    else
      dst_len = x * 10
    end if

    ! Create the output of collect8 and save in compare_dst array
    k = 1

    do i = 0, npes - 1, 1
      do j = 1, mod(i, 4) + 1, 1
        compare_dst4(k) = src4(j)
        k = k + 1
      end do
    end do

    do i = 1, dst_len, 1
      dst4(i) = -1
    end do

    call shmem_barrier_all()

    call shmem_collect8(dst4, src4, mod(me, 4) + 1, &
      0, 0, npes, &
      pSync)

    call shmem_barrier_all()
  
    if(me .eq. 0) then
      do i = 1, dst_len, 1
        if(dst4(i) .ne. compare_dst4(i)) then
          success = 1
        end if
      end do 
      if(success .eq. 1) then
        write (*,*) "Test shmem_collect8: Failed"
      else
        write (*,*) "Test shmem_collect8: Passed"
      end if
    end if
  else
    write (*,*) "Number of PEs must be > 1 to test collects, test skipped"
  end if
end program test_shmem_collects
